File: haskell/day1.hs
import Data.List
import System.IO
import Control.Monad

main = do
    contents <- readFile "input/day1.txt"
    let line_arr = map words (lines contents)
    let list1 = getList line_arr 0
    let list2 = getList line_arr 1
    print (part1 list1 list2)
    print (part2 list1 list2)

getList :: [[String]] -> Int -> [Int]
getList list idx = sort (map (read . (!! idx)) list)

part1 :: [Int] -> [Int] -> Int
part1 list1 list2 = foldl (\acc item -> acc + abs (snd item - list2 !! fst item) ) 0 (zip [0..] list1)

part2 :: [Int] -> [Int] -> Int
part2 list1 list2 = foldl (\acc item -> acc + item * length (filter (== item) list2) ) 0 list1

File: kotlin/app/src/main/kotlin/days/Day1.kt
package days.day1

import java.io.File
import kotlin.math.abs

fun day1() {
    val lines = File("../../input/day1.txt").readText().lines().map { s -> s.split("\\s+".toRegex())}
    val list1 = lines.map{s -> s[0].toInt()}.sorted()
    val list2 = lines.map{s -> s[1].toInt()}.sorted()
    part1(list1, list2)
    part2(list1, list2)
}

fun part1(list1: List<Int>, list2: List<Int>) {
    val sum = list1.withIndex().fold(0) { acc, item: IndexedValue<Int> -> acc + abs(item.value - list2[item.index]) }
    println(sum)
}

fun part2(list1: List<Int>, list2: List<Int>) {
    val sum = list1.fold(0) { acc, item: Int -> acc + item * list2.count{it == item}}
    println(sum)
}